mesh/profiles/views.py

```diff
@@ -44,3 +60,169 @@ def profile_picture(request):
+        POST to upload a profile picture,
+        PATCH to update a profile picture,
+        DELETE to delete a profile picture.
+    """
+    
+    def post(self, request, *args, **kwargs):
+        """
+            Handles POST requests.
+
+            Grabs an existing Profile and uploads a profile picture to imgur,
+            and saves the link with the Profile.
+
+            The required fields are "accountID" and "profilePicture".
+
+            Note: A user should not have a profile picture by this point, as in their
+            profilePicture URL should be empty. If they already have a picture, 
+            a PATCH request should be made instead. If a user wishes to delete their picture,
+            a DELETE request should be made instead.
+        """
+        
+        try:
+            data = request.POST
+
+            account_id = data["accountID"]
+            
+            profile = Profile.objects.get(accountID=account_id)
+            
+            # Check if user already has a profilePicture
+            if profile.profilePicture is not None:
+                return JsonResponse({"error": "Profile already has a profilePicture. " +
+                                     "Use PATCH to update the profilePicture or " + 
+                                     "DELETE to delete the profilePicture."}, 
+                                    status = 409)
+            
+            imgur_client = initialize_imgur_client()
+            
+            # Grab raw photo from request files and encode it to upload to imgur
+            image_file = b64encode(request.FILES["profilePicture"].read())
+
+            # send POST request to imgur API to upload photo
+            imgur_response = imgur_client._send_request('https://api.imgur.com/3/image', 
+                                   method='POST', params={'image': image_file})
+
+            image_link = imgur_response["link"]
+
+            # Save photo URL with user
+            profile.profilePicture = image_link
+            profile.save()
+
+            request_response = json.dumps({"profileID": profile.accountID.accountID ,
+                                 "profilePicture": image_link})
+            
+            return JsonResponse(request_response, status = 201, safe = False)
+
+        except MultiValueDictKeyError:
+            return JsonResponse({"error": "Missing required JSON fields."}, status = 400)
+
+        except Profile.DoesNotExist:
+            return JsonResponse({"error": "Profile not found."}, status = 404)
+        
+        except HTTPError:
+            return JsonResponse({"error": "Uploaded file is not an image."}, status = 400)
+
+        except ValueError:
+            return JsonResponse({"error": "accountID or profilePicture field is empty."}, status = 400)
+
+    def patch(self, request, *args, **kwargs):
+
+        try:
+            data = MultiPartParser(request.META, request, request.upload_handlers).parse()
```

Comment:
Can you explain what this is?
Also, for POST you used request.POST and request.FILES while in PATCH, you used whatever request.META and MultiPartParser
Then you used validate_json_and_required_fields and request.body for the DELETE request

What's the difference between the three?

Reply:
So for PATCH requests, a file will be uploaded, but we can't use request.POST to get the form data, because we aren't receiving a POST request. So a work-around is to use the MutliPartParser that Django provides, as we still need an easy way to grab the files uploaded.

The POST request and PATCH requests both should use forms when sent, because you can't send files over JSON, right? So we need to use request.POST/request.FILES/MultiPartParser to actually access the uploaded files in their respective request methods. 

The DELETE request DOES use JSON, so we can use the validate_json_and_required_fields method, but the other two do NOT use JSON, so we don't have access to that method.